### PR TITLE
10: 1-3

### PR DESCRIPTION
**Pull request by dijeth**

### CODE REVIEW

**Comment by jsru** [comment]
Добавляю преподавателя (@ShGKme) для код-ревью.

File: 10-slots/01-UiPaginationView/components/UiPaginationView.vue [hunk 1]
```diff
@@ -1,10 +1,14 @@
 <template>
   <div class="pagination-container">
-    <!-- Контент страницы -->
+    <template v-for="{ id, item } in pageItems" :key="id">
+      <slot :item="item" />
+    </template>
```
Comment (ShGKme) [inline 10-slots/01-UiPaginationView/components/UiPaginationView.vue]: In Vue 3 `v-for` directive can be directly used on `<slot>`

File: 10-slots/02-PromiseWrapper/components/PromiseWrapper.vue [hunk 2]
```diff
@@ -1,16 +1,71 @@
...
+  state: (typeof StateType)[keyof typeof StateType];
+  result: any;
+  error: Error | null;
+};
+
+export default defineComponent({
   name: 'PromiseWrapper',
 
+  StateType,
+
```
Comment (ShGKme) [inline 10-slots/02-PromiseWrapper/components/PromiseWrapper.vue]: Constants can be used in a component without putting them to options. It is required when we need a constant available in the instance's context for a template.

File: 10-slots/01-UiPaginationView/components/UiPaginationView.vue [hunk 3]
```diff
@@ -26,6 +30,20 @@ export default {
...
     },
   },
+
+  computed: {
+    localItems() {
+      return this.items.map((it) => ({
+        id: --id,
+        item: it,
+      }));
+    },
```
Comment (ShGKme) [inline 10-slots/01-UiPaginationView/components/UiPaginationView.vue]: This is an example of a bad way to add a key. This computed `localItems` will be recreated with a new `id` on any `items` changes. It means, that on any list change, an entire rendered list will be re-created.

Comment (ShGKme) [reply 10-slots/01-UiPaginationView/components/UiPaginationView.vue]: In this component, we have to options:
1. Ask the component's user to provide object property for `key` or function
2. Ask the component's user to manually set `key` in the template provided to the slot

I prefer the second one.

File: 10-slots/03-UiCalendarView/components/MeetupsCalendar.vue [hunk 4]
```diff
@@ -1,16 +1,29 @@
...
+  <UiCalendarView v-slot="{ day, year, month }">
+    <UiCalendarEvent
+      v-for="meetup in getDateMeetups(day, month, year)"
+      tag="a"
+      :href="`/meetups/${meetup.id}`"
+      :key="meetup"
+    >
+      {{ meetup.title }}
     </UiCalendarEvent>
   </UiCalendarView>
```
Comment (ShGKme) [inline 10-slots/03-UiCalendarView/components/MeetupsCalendar.vue]: Calling methods in a Vue template is a bad practice in general. It's ok when it is a fast simple method of output formatting. But `getDateMeetups` is a data processing method. It is moving component's logic, a data processing logic, to the template.

Also, it is not good for performance. On each component re-render, for rendering each cell, it makes a full iteration over the entire array. 

Instead, it's better to prepare data for optimal and simple rendering. For example, creating an object, map from date to meetup on this date.

**Review by ShGKme — CHANGES_REQUESTED** [review]
`10-slots/01-UiPaginationView` has a problem with wrong re-render during to id generation.

**Comment by jsru** [comment]
Добавляю преподавателя (@ShGKme) для код-ревью.

File: 10-slots/02-PromiseWrapper/components/PromiseWrapper.vue (hunk 2 above)
Comment (dijeth) [reply 10-slots/02-PromiseWrapper/components/PromiseWrapper.vue]: yes I forgot to remove it

**Review by ShGKme — APPROVED** [review]

**Pull request merged**